<a href="https://colab.research.google.com/github/Siddhesh-Agarwal/hummingbird-ml-test/blob/main/main.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install wheel

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
!pip install hummingbird-ml[extra]

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Using cached hummingbird_ml-0.4.4-py2.py3-none-any.whl (181 kB)
  Using cached onnxconverter_common-1.12.2-py2.py3-none-any.whl (83 kB)
  Using cached prophet-1.0.1.tar.gz (65 kB)
  Using cached cmdstanpy-0.9.68-py3-none-any.whl (49 kB)
  Using cached pystan-2.19.1.1-cp37-cp37m-manylinux1_x86_64.whl (67.3 MB)
     |████████████████████████████████| 13.1 MB 6.5 MB/s 
  ERROR: Failed building wheel for prophet
  Running setup.py clean for prophet
Failed to build prophet
  Attempting uninstall: pystan
    Found existing installation: pystan 3.3.0
    Uninstalling pystan-3.3.0:
      Successfully uninstalled pystan-3.3.0
  Attempting uninstall: cmdstanpy
    Found existing installation: cmdstanpy 1.0.4
    Uninstalling cmdstanpy-1.0.4:
      Successfully uninstalled cmdstanpy-1.0.4
  Attempting uninstall: prophet
    Found existing installation: prophet 1.1
    Uninstalling prophet-1.1:
   

In [ ]:
import numpy as np
from sklearn.ensemble import RandomForestClassifier
from hummingbird.ml import convert

/usr/local/lib/python3.7/dist-packages/sklearn/experimental/enable_hist_gradient_boosting.py:17: UserWarning: Since version 1.0, it is not needed to import enable_hist_gradient_boosting anymore. HistGradientBoostingClassifier and HistGradientBoostingRegressor are now stable and can be normally imported from sklearn.ensemble.
  "Since version 1.0, "


In [ ]:
num_classes = 2
X = np.random.rand(100000, 30)
y = np.random.randint(num_classes, size=100000)

In [ ]:
model_1 = RandomForestClassifier(n_estimators=10, max_depth=10)

In [ ]:
# time taken by 1st model
%%timeit
model_1.fit(X, y)

5.21 s ± 993 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [ ]:
model_2 = convert(model_1, "pytorch")

/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:103: FutureWarning: Attribute `n_features_` was deprecated in version 1.0 and will be removed in 1.2. Use `n_features_in_` instead.
  warnings.warn(msg, category=FutureWarning)


In [ ]:
# time taken by 2nd model
%%timeit
model_2.predict(X)

143 ms ± 5.45 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [ ]:
model_2.to('cuda')

In [17]:
# time taken by 3rd model
%%timeit
model_2.predict(X)

CPU times: user 16.5 ms, sys: 0 ns, total: 16.5 ms
Wall time: 20.7 ms


array([1, 0, 0, ..., 0, 0, 0])

## Observation
- **model 1**: SciKit Learn model on CPU
  - time: 5.21 s per loop
- **model 2**: PyTorch model on CPU
  - time: 143 ms per loop
- **model 3**: PyTorch model on GPU
  - time: 11.9ms per loop

## Conclusion

- model 2 is 36.43 times faster than model 1
- model 3 is 12.02 times faster than model 2
- model 3 is 437.8 times faster than model 1

In [ ]:
# time taken by model 1 to time taken by model 2
5210 / 143

36.43356643356643

In [ ]:
# time taken by model 2 to time taken by model 3
143/11.9

12.016806722689076

In [ ]:
# time taken by model 1 to time taken by model 3
5210/11.9

437.81512605042013